HAVE_FSTATAT and HAVE_DIRFD are defined using #cmakedefine01, which
means they will always be defined, with a value of 1 if the feature is
available or 0 if it is not. So simply checking if they are defined is
incorrect and leads to the functions being used when unavailable.

--- unix/sftpserver.c.orig	2022-05-25 02:56:28.000000000 +1000
+++ unix/sftpserver.c	2022-06-13 00:27:02.000000000 +1000
@@ -602,7 +602,7 @@ static void uss_readdir(SftpServer *srv,
         char *longnamebuf = NULL;
         struct fxp_attrs attrs = no_attrs;
 
-#if defined HAVE_FSTATAT && defined HAVE_DIRFD
+#if HAVE_FSTATAT && HAVE_DIRFD
         struct stat st;
         if (!fstatat(dirfd(udh->dp), de->d_name, &st, AT_SYMLINK_NOFOLLOW)) {
             char perms[11], *uidbuf = NULL, *gidbuf = NULL;
